### PR TITLE
Only `SetTail` for Accepted Blocks

### PR DESCRIPTION
**Pull request by patrick-ogrady**
This PR alters the `vm` logic to only `SetTail` when a block has been accepted. To avoid creating additional contention on block production, this PR alters the `miner` to build off of the last preferred block (instead of the last block we called `SetTail` on).

### CODE REVIEW

File: plugin/evm/vm.go [hunk 1]
```diff
@@ -513,8 +514,13 @@ func (vm *VM) GetBlock(id ids.ID) (snowman.Block, error) {
 
 // SetPreference sets what the current tail of the chain is
 func (vm *VM) SetPreference(blkID ids.ID) {
-	err := vm.chain.SetTail(common.Hash(blkID))
-	vm.ctx.Log.AssertNoError(err)
+	block := vm.getBlock(blkID)
+	if block == nil {
+		vm.ctx.Log.Error("could not get preference block %s", blkID)
```
Comment (aaronbuchwald) [inline plugin/evm/vm.go:519]: Could you change this to `vm.ctx.Log.AssertNoError(err)`? I think we should treat this as a critical error

**Review by StephenButtolph — APPROVED** [review]

File: core/blockchain.go [hunk 2]
```diff
@@ -857,6 +858,109 @@ func (bc *BlockChain) GetBlockByNumber(number uint64) *types.Block {
...
+	if currentSize > 0 {
+		bc.chainmu.Lock()
+		// Flush the whole batch into the disk, exit the node if failed
+		if err := batch.Write(); err != nil {
+			bc.chainmu.Unlock()
+			log.Crit("Failed to update chain indexes and markers", "err", err)
+		}
+		bc.chainmu.Unlock()
+		totalUpdates += currentSize
+		log.Info("canonical chain update", "totalUpdates", totalUpdates)
```
Comment (patrick-ogrady) [inline core/blockchain.go:958]: let's break this out of this if to always print total updates

File: core/blockchain.go [hunk 3]
```diff
@@ -857,6 +858,109 @@ func (bc *BlockChain) GetBlockByNumber(number uint64) *types.Block {
...
+			return fmt.Errorf("failed to get block by number at height: %d", blkNumber)
+		}
+		if canonicalBlk.Hash() == current.Hash() {
+			continue
+		}
+
+		// If the canonical blockhash at [blkNumber] is incorrect
+		// repair it here.
+		rawdb.WriteCanonicalHash(batch, current.Hash(), current.NumberU64())
+		rawdb.WriteTxLookupEntriesByBlock(batch, current)
```
Comment (patrick-ogrady) [inline core/blockchain.go:933]: we should log each repair after this line

File: core/blockchain.go [hunk 4]
```diff
@@ -857,6 +858,109 @@ func (bc *BlockChain) GetBlockByNumber(number uint64) *types.Block {
...
+			bc.chainmu.Lock()
+			// Flush the whole batch into the disk, exit the node if failed
+			if err := batch.Write(); err != nil {
+				bc.chainmu.Unlock()
+				return fmt.Errorf("failed to write batch with size %d at current block height %d: %s", currentSize, current.NumberU64(), err)
+			}
+			bc.chainmu.Unlock()
+			totalUpdates += currentSize
+			currentSize = 0
+			log.Info("canonical chain update", "currentBlock", current.NumberU64(), "totalUpdates", totalUpdates)
```
Comment (patrick-ogrady) [inline core/blockchain.go:945]: let's change this to write batch

**Pull request merged**